File: exercise/ts/day24/src/domain/StockUnit.ts
import {Either, left, right} from 'fp-ts/lib/Either';
import {DomainError} from "./core/DomainError";

export class StockUnit {
    constructor(public value: number) {
    }

    static from(stock: number): Either<DomainError, StockUnit> {
        return stock >= 0
            ? right(new StockUnit(stock))
            : left(new DomainError('A stock unit cannot be negative'));
    }

    isSupplied(): boolean {
        return this.value > 0;
    }

    decrease(): StockUnit {
        return new StockUnit(this.value - 1);
    }
}

File: exercise/java/day24/src/main/java/domain/StockUnit.java
package domain;

import io.vavr.control.Either;

import static io.vavr.control.Either.left;
import static io.vavr.control.Either.right;

public record StockUnit(int value) {

    public static Either<Error, StockUnit> from(int stock) {
        return stock >= 0
                ? right(new StockUnit(stock))
                : left(new Error("A stock unit cannot be negative"));
    }

    public boolean isSupplied() {
        return value > 0;
    }

    public StockUnit decrease() {
        return new StockUnit(value - 1);
    }
}